astroid/brain/brain_typing.py

```diff
@@ -344,6 +346,34 @@ def infer_tuple_alias(
     return iter([class_def])
 
 
+def _looks_like_typing_cast(node: Call) -> bool:
+    return isinstance(node, Call) and (
+        isinstance(node.func, Name)
+        and node.func.name == "cast"
+        or isinstance(node.func, Attribute)
+        and node.func.attrname == "cast"
+    )
+
+
+def infer_typing_cast(
+    node: Call, ctx: context.InferenceContext = None
+) -> typing.Iterator[Instance]:
+    """Infer call to cast() returning an instance of the cast-to type"""
+    try:
+        func = next(node.func.infer(context=ctx))
+    except InferenceError as exc:
+        raise UseInferenceDefault from exc
+    if func.qname() != "typing.cast" or len(node.args) != 2:
+        raise UseInferenceDefault
+
+    peek_value_node, value_node = itertools.tee(node.args[1].infer(context=ctx))
```

Comment:
The use of `itertools.tee` isn't really necessary. We are only really interested in the first return value.


This should be enough
```py
    inferred = next(node.args[1].infer(context=ctx)
```

Reply:
My first attempt was roughly what you describe here, but it didn't work, because if the first element of the iterable isn't uninferable then we want to return the whole iterator (so all values inferred for `x` can be inferred as the return type of `cast()`. AIUI by having called `next()` I've already consumed the first element of the iterator at this point, so I'm not able to return the full iterator.

Am I understanding this right?

Reply:
Though about this again. Yeah, you are right. I didn't account for the case that more than one inference result would be returned (I believe in most situations it's only one).

To still avoid the overhead of `itertools.tee`, you could use one of the following options:
1. Store results of `infer` in list and check against first value / return `iter(inferred_lst)`.
2. Just call `node.args[1].infer(context=ctx)` two times. Inference results are already cached with `context`, so I suspect there isn't much / if any performance hit.
https://github.com/PyCQA/astroid/blob/c9c498348174b38ce35bfe001353c8ebea262802/astroid/node_classes.py#L357-L360

Reply:
I believe the `tee()` solution has less overhead than either of your suggestions, not more. It guarantees that the original generator will be executed at most once, and has only trivial temporary storage in this case (where only one element is being stored in a temporary FIFO, which will presumably be consumed the moment we return our result to the caller).

Never the less, if there's no significant overhead to running the inference twice then that has the advantage of making the intent of the code much clearer, so I've done that. Thanks for the suggestion.

Reply:
I don't fully understand what's going on here. If I call the `node.args[1].infer(context=ctx)` twice then it doesn't work the second time. Trying to consume it gives:

    astroid.exceptions.InferenceError: StopIteration raised without any error information.

If I don't pass in the context the second time then it works as expected, so presumably there's some state being stored in the context that is making the second inference depend on the first. I don't really understand what's stored in the context and if this is expected behaviour, though.

Reply:
Found the issue: The inference result is only cached in the context if the first iterator is run till completion, i.e. `tuple(node.args[1].infer(context=ctx))[0] is Uninferable`. Without that the context only knows that it has seen the name for `args[1]` already and thus thinks it can safely abort by raising a `StopIteration`.

You made a good point regarding `itertools.tee()`, so let's use it then. Thank you for trying it though.

Reply:
Thanks for the clear explanation of what's happening. Saved me a few hours' debugging!